### PR TITLE
Fix byte order for scripts and endianness for Hash <> U256 conversion

### PR DESCRIPTION
**Pull request by Jeanmichel7**
- [x] issue #101 
- [x] follows contribution [guide](https://github.com/keep-starknet-strange/raito/blob/main/CONTRIBUTING.md)
- [x] code change includes tests

### CODE REVIEW

File: src/utils/hex.cairo [hunk 1]
```diff
@@ -34,6 +34,27 @@ fn hex_char_to_nibble(hex_char: u8) -> u8 {
     }
 }
 
+/// Print a byte array as hex
+pub fn print_bytearray_hex(data: @ByteArray) {
```
Comment (m-kus) [inline src/utils/hex.cairo:38]: If you want to keep is a part of this PR i'd suggest to refactor this as `to_hex(data: @ByteArray) -> ByteArray` and add a test.

The you can use it as `println!("{}", to_hex(...))` which is more flexible and intuitive

File: src/utils/hash.cairo [hunk 2]
```diff
@@ -88,28 +88,41 @@ pub impl U256IntoHash of Into<u256, Hash> {
...
+/// Hash is little-endian order, while u256 is big-endian like in explorer.
 pub impl HashIntoU256 of Into<Hash, u256> {
     fn into(self: Hash) -> u256 {
-        let [a, b, c, d, e, f, g, h] = self.value;
+        let [mut a, mut b, mut c, mut d, mut e, mut f, mut g, mut h] = self.value;
 
         let mut low: u128 = 0;
         let mut high: u128 = 0;
 
-        low += (h.into());
```
Comment (m-kus) [inline src/utils/hash.cairo:99]: Please keep it simple, it's much more readable this way, also avoid mutable variables if possible

File: src/utils/hash.cairo [hunk 3]
```diff
@@ -88,28 +88,41 @@ pub impl U256IntoHash of Into<u256, Hash> {
     }
 }
 
-/// Converts a `Hash` value into a `u256` type.
+/// Converts a `Hash` value into a `u256` type and reverse bytes order.
+/// Hash is little-endian order, while u256 is big-endian like in explorer.
 pub impl HashIntoU256 of Into<Hash, u256> {
```
Comment (m-kus) [inline src/utils/hash.cairo:93]: U256IntoHash also has to be updated

File: tests/blocks/block_170.cairo [hunk 4]
```diff
@@ -54,8 +54,9 @@ pub fn block_170() -> Block {
                         sequence: 4294967295,
                         witness: array![].span(),
                         previous_output: OutPoint {
-                            txid: 0x0437cd7f8525ceed2324359c2d0ba26006d92d856a9c20fa0241106ee5a597c9_u256
```
Comment (m-kus) [inline tests/blocks/block_170.cairo:57]: This was correct, reversed hex coming from RPC is equivalent to little endian encoding of U256

File: src/validation/work.cairo [hunk 5]
```diff
@@ -33,23 +33,33 @@ mod tests {
     #[test]
     fn test_validate_proof_of_work() {
         // target is less than prev block hash
-        let result = validate_proof_of_work(0, 1_u256.into());
```
Comment (m-kus) [inline src/validation/work.cairo:36]: This was correct

File: src/validation/work.cairo [hunk 6]
```diff
@@ -33,23 +33,33 @@ mod tests {
...
-        let result = validate_proof_of_work(1, 2_u256.into());
+        let result = validate_proof_of_work(
+            1, 0x0200000000000000000000000000000000000000000000000000000000000000_u256.into()
+        );
         assert!(result.is_err(), "Expect prev block hash gt target");
 
         // block prev block hash is less than target
-        let result = validate_proof_of_work(10, 9_u256.into());
+        let result = validate_proof_of_work(
+            10, 0x0900000000000000000000000000000000000000000000000000000000000000_u256.into()
```
Comment (m-kus) [inline src/validation/work.cairo:61]: Even intuitively, 9 < 10

File: src/utils/hash.cairo [hunk 7]
```diff
@@ -88,28 +88,41 @@ pub impl U256IntoHash of Into<u256, Hash> {
...
+        if (!is_last || i != 3) {
+            result = shl(result, 8_u128);
+            value = shr(value, 8_u32);
+        }
+        i += 1;
+    };
+}
+
 #[cfg(test)]
 mod tests {
```
Comment (m-kus) [inline src/utils/hash.cairo:127]: Please add a test where you do a round trip, i.e. cast U256 to Hash and back

File: src/utils/hash.cairo (hunk 2 above)
Comment (Jeanmichel7) [reply src/utils/hash.cairo:99]: Like that ?

Comment (m-kus) [reply src/utils/hash.cairo:99]: I think you can just use `u128_byte_reverse` function from `core::integer`

**Pull request merged**